keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,386 @@
+
+A natural question is how we should define the matching behavior. Suppose there are two ResourceClass objects. ResourceClass RC1 has metadata matching constraint “property1 = value1”, and ResourceClass RC2 has metadata matching constraint “property2 = value2”. Suppose a ComputeResource has both “property1: value1” and “property2: value2” properties, and so match both ResourceClasses. Now should the scheduler consider this ComputeResource as qualified for both ResourceClasses RC1 and RC2, or only one of them?
+
+We feel the desired answer to this question may vary across different types of resources, properties and use cases. To illustrate this, lets consider the following example: A GPU device plugin in a cluster with different types of GPUs may be configured to advertise under a common ResourceName as "nvidia.com/nvidia-tesla-k80", at the beginning. To support per GPU type resource quota, cluster admins may define the following ResourceClasses:
+
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia-k80
+spec:
+  resourceName: "nvidia.com/nvidia-tesla-k80"
+
+kind: ResourceClass
+metadata:
+  name: nvidia-p100
+spec:
+  resourceName: "nvidia.com/nvidia-tesla-p100"
+```
+
+Later on, suppose the cluster admins add a new GPU node group with a new version of GPU device plugin that exports another resource property "Nvlink" which will be set true for nvidia-tesla-p100 GPUs connected through nvlinks. To utilize this new feature, the cluster admins define the following new ResourceClass with nvlink constraints:
+
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia-p100-nvlink
+spec:
+  resourceName: "nvidia.com/nvidia-tesla-p100"
+  labelSelector:
+        - key: "Nvlink"
+          operator: "Eq"
+          values:
+            - "true"
+```
+
+Now we face the question that whether the scheduler should allow Pods requesting
+"nvidia-p100" to land on a node in this new GPU node groups. So far, we have
+received different feedbacks on this question. In some use cases, users would
+like to have minimum matching behavior that as long as the underlying hardware
+matches the minimal requirements specified through ResourceClass contraints,
+they want to allow Pods to be scheduled on the hardware. On the other hand, some users desire to reserve expensive hardware resources for users who explicitly request them.
+We feel both use cases are valid requirements. Allowing a ComputeResource to match
+multiple ResourceClasses as long as it matches their matching constraints
+perhaps yields least surprising behavior to users and also simplies upgrade
+scenario as new resource properties are introduced into the system. Therefore we
+support this behavior by default. To also provide an easy way for cluster admins
+to reserve expensive compute resources and control their access with resource
+quota, we propose to include a Priority field in ResourceClass API.
```

Comment:
Could we clarify the use cases that require non-overlapping resource classes?

1. It seems this effect is achievable anyway if cluster admins design their resource class specs properly.
1. With the described priority mechanism, to answer the question "why doesn't resource X on node Y match class Z?" users potentially have to inspect every resource class.